proxy/threads.c

```diff
@@ -0,0 +1,176 @@
+#define _GNU_SOURCE
+
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <memory.h>
+#include <pthread.h>
+
+
+#include "queue.h"
+#include "sockets.h"
+#include "mypoll.h"
+#include "threads.h"
+
+#define BUFFER_SIZE 4096
+
+
+
+void* connectionThread(void* arg) {
+    Data d = *(Data*)arg;
+
+    queue_t* newConnectQueue = d.queue;
+    sPoll* spoll = d.poll;
+    //ssize_t bytes_read;
+
+    for (;;) {
+        Sockets* newSockets;
+        queue_get(newConnectQueue, (void**)&newSockets);
+        char buffer[BUFFER_SIZE];
+        // Чтение запроса от клиента
+        int readed;
+        //printf("начал читать %d\n", newSockets->client);
+        if (( readed = read(newSockets->client, buffer, BUFFER_SIZE)) > 0) {
+        }
+
+        if (readed == -1) {
+            printf("соединение разорвано\n");
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;
+        }
+
+        int dnslen;
+        char *start, *end;
+
+        start = strstr(buffer, "//");
+        if (!start) {
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;
+        }
+        start+=2;
+
+        end = strchr(start, '/');
+        if (!end) {
+            closeSockets(*newSockets);
+            free(newSockets);
+            continue;     
+        }
+        dnslen = end-start;
+
+        
+
+        char* dns = (char*)malloc(dnslen+1);
+        strncpy(dns, start, dnslen);
+        
+
+
+        const char *port = "80";
+        const char *path = "/";
+        // Получаем IP-адрес по доменному имени
+        struct addrinfo hints, *result, *rp;
+        memset(&hints, 0, sizeof(struct addrinfo));
+        hints.ai_family = AF_UNSPEC;
+        hints.ai_socktype = SOCK_STREAM;
+    
+        int status = getaddrinfo(dns, port, &hints, &result);
+        if (status != 0) {
+            fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(status));
+            close(newSockets->client);
+            continue;;
+        }
+
+        // Перебираем результаты, используем первый подходящий IP-адрес
+        int soc = -1;
+        for (rp = result; rp != NULL; rp = rp->ai_next) {
+            soc = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
```

Comment:
Please list all possible types of sockets and families. Are all of them available for HTTP communication ?

Reply:
?